### PR TITLE
migrate spice to npm and upgrade

### PR DESCRIPTION
**Pull request by ohadlevy**
* move from an old version of spice to a much newer one (over 5 years upgrade)
* move spice js to webpack

### CODE REVIEW

File: webpack/assets/javascripts/spice.js [hunk 1]
```diff
@@ -0,0 +1,75 @@
+import { SpiceMainConn } from '@spice-project/spice-html5';
+import { sprintf, translate as __ } from './react_app/common/I18n';
+
+export { sendCtrlAltDel } from '@spice-project/spice-html5';
+
+let sc = null;
+
+export function startSpice() {
+  const scheme = 'ws://';
```
Comment (ekohl) [inline webpack/assets/javascripts/spice.js:10]: So no support for `wss://`?

File: app/views/hosts/console/spice.html.erb [hunk 2]
```diff
@@ -18,3 +18,5 @@
   </div>
   <div id="spice-screen" class="console-screen"> </div>
 <% end %>
+
+<script type="text/javascript"> tfm.spice.startSpice() </script>
```
Comment (ekohl) [inline app/views/hosts/console/spice.html.erb]: nit: newline

File: webpack/assets/javascripts/spice.js (hunk 1 above)
Comment (ekohl) [reply webpack/assets/javascripts/spice.js:10]: Looks like I forgot to hit save on my update. I now see the old code was `ws://` only as well. Considering browsers disallow mixed content by default I wonder how many people actually use this.

Comment (ohadlevy) [reply webpack/assets/javascripts/spice.js:10]: I'm not changing anything we didn't have before :)

Comment (ohadlevy) [reply webpack/assets/javascripts/spice.js:10]: > Looks like I forgot to hit save on my update. I now see the old code was `ws://` only as well. Considering browsers disallow mixed content by default I wonder how many people actually use this.

TBH I tend to agree, from very quick testing, it doesn't seems to work on libvirt at all? and could only work with oVirt. I'm fine to suggest removing it from the code base as well....

**Comment by ohadlevy** [comment]
@UXabre any chance you can test this out? thanks!

**Comment by UXabre** [comment]
Awesome! Will test this tomorrow @ohadlevy! Was also looking into this one last year but didn't take the time :-( will keep you posted

**Comment by UXabre** [comment]
Hey @ohadlevy, tested this patch and it seems to work as expected. Had hoped that some of the issues had been resolved but seems there's still some work that could be done; at least the deploy flow makes more sense and will make it easier to patch stuff and get them into Foreman (IMHO).

Great work!

**Comment by UXabre** [comment]
Ow, wait, perhaps I've been a bit to quick on this one; just tested this out on both a windows machine and a linux machine but it seems that the CTRL+ALT+DEL button does not work anymore and an error pops up:
```

inputs.js:195 Uncaught ReferenceError: sc is not defined
    at Object.sendCtrlAltDel (inputs.js:195)
    at HTMLAnchorElement.onclick (VM103 console:81)

```

Also, in firefox I noticed an extra button for "New Window" (first time I noticed this one actually :p) but it also does not work, stating:

`TypeError: pluginobj.connect is not a function bundle.js:32731:3`

Which in code seems to point to this section of code:

```
function connectXPI() {
  if ($('#spice-xpi').length === 0) {
    $('#spice-area').append('<embed type="application/x-spice" height=0 width=0 id="spice-xpi">');
  }
  var attrs = $('#spice-area');
  // we close down the other WebSocket connection when opening the XPI
  disconnect();
  var pluginobj = document.embeds[0];
  pluginobj.hostIP = attrs.data('address');
  pluginobj.SecurePort = attrs.data('secure-port');
  pluginobj.Password = attrs.data('password');
  pluginobj.TrustStore = decodeURIComponent(attrs.data('ca-cert'));
  pluginobj.SSLChannels = String('all');
  pluginobj.fullScreen = false;
  pluginobj.Title = attrs.data('title');
  pluginobj.HostSubject = attrs.data('subject');
  pluginobj.connect();
}

```

On a positive note, windows is drawn correctly! Before it had this upside down behaviour but it seems to be fixed so hooray for progress! :-D


Kind regards,
Arend

File: webpack/assets/javascripts/spice.js (hunk 1 above)
Comment (ohadlevy) [reply webpack/assets/javascripts/spice.js:10]: @ekohl for the record - https://projects.theforeman.org/issues/22801

**Comment by ohadlevy** [comment]
> Ow, wait, perhaps I've been a bit to quick on this one; just tested this out on both a windows machine and a linux machine but it seems that the CTRL+ALT+DEL button does not work anymore and an error pops up:

thanks, I believe I fixed that.



> Also, in firefox I noticed an extra button for "New Window" (first time I noticed this one actually :p) but it also does not work, stating:
> 
> `TypeError: pluginobj.connect is not a function bundle.js:32731:3`
> 
> Which in code seems to point to this section of code:
> 
> ```
> function connectXPI() {
>   if ($('#spice-xpi').length === 0) {
>     $('#spice-area').append('<embed type="application/x-spice" height=0 width=0 id="spice-xpi">');
>   }
>   var attrs = $('#spice-area');
>   // we close down the other WebSocket connection when opening the XPI
>   disconnect();
>   var pluginobj = document.embeds[0];
this seems to be the core issue (pluginobj is not what we expect), looking at the code changes over the time, it seems there is no change, can you confirm its broken without the patch as well?

AFAIU - since this was not introduced by this PR, I suggest to not block merging this PR and could be fixed in a follow up PR (that and of course WSS support in https://projects.theforeman.org/issues/22801)

**Comment by ohadlevy** [comment]
@UXabre any chance you can review again? thanks

**Comment by ohadlevy** [comment]
packaging pr at https://github.com/theforeman/foreman-packaging/pull/3691

**Comment by UXabre** [comment]
Hey @ohadlevy , checked it in the (almost) latest version of foreman (1.22.2) and it seems the "New Window" button is not there in firefox (which is strange):
image
In your latest patch I still see this button; but, as you said, perphaps something else is going on here
image

Anyway, just checked out your adjustements and now I'm able to press the ctrl+alt+delete button and it works as expected.

Just curious, does your system also hide the cursor? Currently I can't see any cursor on screen BUT I do have a touch-screen enabled computer (although I never really use it), so maybe it tries to be clever and hides the cursor on touch-screen enabled machines?

**Comment by mmoll** [comment]
@ohadlevy please rebase :)

**Comment by ohadlevy** [comment]
rebased.

**Comment by mmoll** [comment]
@UXabre just to make sure: the non-displaying button is not caused by this patch, correct?

**Comment by UXabre** [comment]
> @UXabre just to make sure: the non-displaying button is not caused by this patch, correct?

Hello, actually, the button only started showing on Firefox **since** this patch. 
Looking at the code, seems the "issue" originates here (but as far as I understood, this code was always there, but just refactored a bit now...although, I don't have this button on other versions of foreman):

https://github.com/theforeman/foreman/pull/6701/files#diff-2babf5acd1a15f527021c7a01aa4a38fR5

**Review by mmoll — APPROVED** [review]

**Comment by mmoll** [comment]
merged, thanks @ohadlevy!

File: webpack/assets/javascripts/spice.js (hunk 1 above)
Comment (ekohl) [reply webpack/assets/javascripts/spice.js:10]: And completing it: https://github.com/theforeman/foreman/pull/8515. It really was adding 2 lines of code.

**Pull request merged**